src/linker/Linker.Dataflow/DynamicallyAccessedMembersTypeHierarchy.cs

```diff
@@ -91,6 +91,11 @@ public DynamicallyAccessedMembersTypeHierarchy (LinkContext context, MarkStep ma
 
 			Debug.Assert (!apply || annotation != DynamicallyAccessedMemberTypes.None);
 
+			// If OptimizeTypeHierarchyAnnotations is disabled, we will apply the annotations without seeing object.GetType()
+			// Unfortunately, we cannot apply the annotation to type derived from EventSource - Revisit after https://github.com/dotnet/runtime/issues/54859
+			apply |= (annotation != DynamicallyAccessedMemberTypes.None) && !_context.IsOptimizationEnabled (CodeOptimizations.OptimizeTypeHierarchyAnnotations, type)
+				&& !_context.IsOptimizationEnabled (CodeOptimizations.RemoveEventSourceSpecialHandling, type) && !BCL.EventTracingForWindows.IsEventSourceImplementation (type, _context);
```

Comment:
Yes, the hack to avoid special handling of `EventSource` is not easy to follow (wish the library source gets fixed so we can avoid this). We wish to apply only if the optimization mode, `OptimizeTypeHierarchyAnnotations`, is disabled. The last 2 are to ensure that `EventSource` types don't follow this.

Reply:
I think it should be
```suggestion
				&& !(_context.IsOptimizationEnabled (CodeOptimizations.RemoveEventSourceSpecialHandling, type) && BCL.EventTracingForWindows.IsEventSourceImplementation (type, _context));
```
so that if we are removing event source handling, it'll still apply the annotations to non-eventsource types.

Reply:
Based on discussions it sounds like I misunderstood the intention - `RemoveEventSourceSpecialHandling == false` means "treat EventSource specially, including by not applying DAMT type hierarchy annotations". In that case I think the logic should be:
```suggestion
				&& (_context.IsOptimizationEnabled (CodeOptimizations.RemoveEventSourceSpecialHandling, type) || !BCL.EventTracingForWindows.IsEventSourceImplementation (type, _context));
```

And a better name might be "DisableEventSourceDiscovery" or "DisableEventSourceSpecialHandling" - how do those sound to you?